src/ProjectManager.cpp

```diff
@@ -530,21 +530,20 @@ void ProjectManager::OnCloseWindow(wxCloseEvent & event)
    // TODO: Is there a Mac issue here??
    // SetMenuBar(NULL);
 
-   // Compact the project.
-   projectFileManager.CompactProjectOnClose();
-
    // Set (or not) the bypass flag to indicate that deletes that would happen during
    // the UndoManager::ClearStates() below are not necessary.
    projectFileIO.SetBypass();
 
-   {
-      // This can reduce reference counts of sample blocks in the project's
-      // tracks.
-      UndoManager::Get( project ).ClearStates();
+   // This can reduce reference counts of sample blocks in the project's
+   // tracks. Do this before `CompactProjectOnClose` so that changes after the
+   // last save don't get written to the project file.
+   UndoManager::Get(project).ClearStates();
 
-      // Delete all the tracks to free up memory
-      tracks.Clear();
-   }
+   // Compact the project.
+   projectFileManager.CompactProjectOnClose();
```

Comment:
In `CompactProjectOnClose` consider `ProjectFileIO::WasCompacted` returns `false`, then it would never execute `ProjectFileIO::UpdateSaved` because undo stack was cleared. I don't know if that was a workaround for some problem, but now it becomes a dead `if`-branch.
Alternatively I could propose to fix `ProjecFileIO::Compact`:
```      
if(!ShouldCompact(tracks))
{
   mWasCompacted = true;
   ...
```
though `ProjectFileIO::SetBypass()` might need changes too.

Reply:
Thanks for this comment, I admit I hadn't examined possible consequences of my change exhaustively.

That dead branch updates the "document blob" (does this include the database?), discarding the unsaved tracks, but doing so it also saves the tempo and time signature changes.

If you change something else not track-related, like the snapping precision in project A, next time you open project B it will have that precision. Something similar happens for tempo, but that's not the desired behaviour, and I guess it's the first thing that's not track-related that must be project-specific.

If that theory is correct, then it'd explain why up to now `projectFileIO.UpdateSaved( mLastSavedTracks.get() );` was ok until time signature was introduced. Follow-up question is whether the change of this commit would have been a correct alternative all this time and this now dead branch can be deleted.

Reply:
In any case, I realize that this commit alone can't be completely correct. `ProjectFileIO::SetBypass()` depends on the execution of `ProjectFileIO::Compact()` (called from `ProjectFileManager::CompactProjectOnClose()`). Just pushed something to fix this.